R/openapi-types.R

```diff
@@ -70,6 +70,24 @@ add_api_info_onLoad <- function() {
     as.character,
     location = c("query", "path")
   )
+  addApiInfo(
+    "date-time",
+    c("POSIXct", "POSIXt"),
```

Comment:
This would allow for people to use 
```r
#* @param x:POSIXct Description of my date `x`
```

Since we are not doing anything in particular about the date type object, let's rearrange the values.

```suggestion
    c("datetime", "date-time"), # Keep `datetime` for convenience
```

Reply:
I've got to admit I don't really follow what the current parameters for addApiInfo mean... and I think #889 might have changed what they mean...

As far as I can see, if we are heading towards auto-conversion by default, then in an apiInfo there should be:

- the type key for the entry - this will be what appears next to the `@param` name - e.g. `@param my_field:int`
- the locations this can be found (I guess... although I've not used this much)
- the openapi type that will be displayed (one of https://swagger.io/docs/specification/data-models/data-types/ ?)
- (optional) the `format` specifier to go alongside the openapi type
- (optional) the R `parser` that will be applied to parse the incoming `character` value (this could possibly also return errors if parsing unsuccessful?)
- (optional) the regex that is applied in the swagger ui
- (optional) possible other keywords like `minimum`, `multipleOf`, 'minLength`, etc (but think they live in a future PR 👍 )

 Based on my understanding from this list... I don't really understand what `apiType`, `plumberTypes` and `realType` are
 - if doing a major release, can we get rid of (or deprecate) some of these fields?
 
Would it help to clean up `addApiInfo` and replace it with a function with a surface something like:
`
addApiInfo(
    key, 
    locations = c("body", "route", "query"), 
    openApiType = c("string", "number", "integer", "boolean", "object"),
    # note that openApiFormat is extensible - so match.arg should not be used on openApiFormat
    openApiFormat = c("float", "double", "int32", "int64", "date", "date-time", "password", "byte", "binary"), 
    openApiRegex  = NULL,
    parser = function(input) { list(error=NULL, result=input) } 
 ) {
 ...
 }
 `
 
That could then be called and configured with only one key for each entry (I think maybe plumber doesn't need this `for` loop any more - https://github.com/rstudio/plumber/blob/befc866cb0d9d24cedf8399c82305c2e2e306214/R/openapi-types.R#L34 ) 

Disclaimer: this is all coming very much from #889 and I admit I don't have any experience of the attachment/file and object inputs... they might cause more problems here?

Sorry... didn't mean to ask such "big questions"

Reply:
I had a go at something like this on a different branch - https://github.com/slodge/plumber/commit/536dd297cc5a8ebef4bb6558e1b044091b0f2d7c

To keep backward portability, I kept all the synonyms - e.g. dbl and double

That change is not complete... several more bits to do including the file/object stuff - but wanted to send it in for discussion... Is this direction worth pursuing or is it too big a change for the core team? 

It's not something I feel we desparately need - it is just a cleanup of api spec definitions rather than anything functional...

Reply:
Much better variable names! Makes it much clearer as to what is happening. I'll admit it was fussy when I wrote it.

Nice work on the backwards compatibility. 

Would be happy to merge the changes once complete.

Reply:
Thanks.

Have got all tests passing (first half of Argentina vs Australia wasn't exciting so I had spare time!)

Have merged those changes into this PR.

Would really like to drop `converterArray` depending on decisions here and in the other PR - https://github.com/rstudio/plumber/pull/889#discussion_r1038748298 - I guess we really only intend this ever to be used for `[int]` and in query paths...

I can merge the PRs into one big thing to review if that would make it easier?

Reply:
Still open here:

- Add the date/datetime formats to annotations.Rmd's table in More details on `@param`  (and more to discuss?)
- Add a NEWS entry ending in (@slodge #891)
- Discuss/resolve this `converterArray` stuff